pkg/job/associator/associator.go

```diff
@@ -0,0 +1,150 @@
+			names[i] = strings.ReplaceAll(names[i], "_", " ")
+			dimensionNames = append(dimensionNames, names[i])
+		}
+		m.dimensions = dimensionNames
+
+		for idx, r := range resources {
+			if mappedResources[idx] {
+				continue
+			}
+
+			match := regex.FindStringSubmatch(r.ARN)
+			if match == nil {
+				continue
+			}
+
+			labels := make(map[string]string, len(match))
+			for i := 1; i < len(match); i++ {
+				labels[names[i]] = match[i]
+			}
+			signature := promCommon.LabelsToSignature(labels)
+			m.mapping[signature] = r
+			mappedResources[idx] = true
+		}
+
+		assoc.mappings = append(assoc.mappings, m)
+	}
+
+	// sort all mappings by decreasing number of dimensions names
+	// (this is essential so that during matching we try to find the metric
+	// with the most specific set of dimensions)
+	slices.SortFunc(assoc.mappings, func(a, b *dimensionsRegexpMapping) bool {
+		return len(a.dimensions) >= len(b.dimensions)
+	})
+
+	return assoc
+}
+
+// AssociateMetricsToResources finds the resource that corresponds to the given set of dimensions
+// names and values of a metric. The guess is based on the mapping built from dimensions regexps.
+// In case a map can't be found, the second return parameter indicates whether the metric should be
+// ignored or not.
+func (assoc Associator) AssociateMetricsToResources(cwMetric *cloudwatch.Metric) (*model.TaggedResource, bool) {
+	if len(cwMetric.Dimensions) == 0 {
+		// Do not skip the metric (create a "global" metric)
+		return nil, false
+	}
+
+	dimensions := make([]string, 0, len(cwMetric.Dimensions))
+	for _, dimension := range cwMetric.Dimensions {
+		dimensions = append(dimensions, *dimension.Name)
+	}
+
+	// Find the mapping which contains the most
+	// (but not necessarily all) the metric's dimensions names.
+	// Mappings are sorted by decreasing number of dimensions names.
+	var regexpMapping *dimensionsRegexpMapping
+	for _, m := range assoc.mappings {
+		if containsAll(dimensions, m.dimensions) {
+			regexpMapping = m
+			break
+		}
+	}
+
+	if regexpMapping == nil {
+		// if no mapping is found, it means the ListMetrics API response
+		// did not contain any subset of the metric's dimensions names.
+		// Do not skip the metric though (create a "global" metric).
+		return nil, false
+	}
+
+	// A mapping has been found. The metric has all (and possibly more)
+	// the dimensions computed for the mapping. Pick only exactly
+	// the dimensions of the mapping to build a labels signature.
+	labels := make(map[string]string, len(cwMetric.Dimensions))
+	for _, rDimension := range regexpMapping.dimensions {
+		for _, mDimension := range cwMetric.Dimensions {
+			if rDimension == *mDimension.Name {
+				labels[*mDimension.Name] = *mDimension.Value
+			}
+		}
+	}
+	signature := promCommon.LabelsToSignature(labels)
+
+	if resource, ok := regexpMapping.mapping[signature]; ok {
+		return resource, false
+	}
+
+	// if there's no mapping entry for this resource, skip it
+	return nil, true
```

Comment:
Getting here means that the labelset mapping exists, but names and values don't match with what we're looking for. That's why I've written "mapping entry".

Maybe the term mapping is confusing in this context.

An example is that we're trying to match dimensions `{a:b, c:d}`. We found a `mapping` (`dimensionsRegexpMapping`) that, say, maps the dimension `a` (a `dimensionsRegexpMapping` might map only a subset of the dimensions we're searching for but still be correct -- because the metric might have more dimensions). So `regexpMapping.mapping` might contain `a:blah` but not `a:b`. This is what I was trying to describe with "no mapping entry found".

Reply:
Ah that makes sense, I was originally confused this comment, https://github.com/nerdswords/yet-another-cloudwatch-exporter/blob/4b6a43d5696ddf8dbb75b896bfebe00d113791cf/pkg/job/maxdimassociator/associator.go#L121, says a mapping has been found but then this one says there's no mapping which seems to contradict the first one.